include/libhal-display/apa102.hpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <array>
+
+#include <libhal-soft/inert_drivers/inert_output_pin.hpp>
+#include <libhal-util/output_pin.hpp>
+#include <libhal-util/spi.hpp>
+
+namespace hal::display {
+/**
+ * @brief Contains data to send over SPI and information about size of the data
+ * to send over
+ *
+ * @tparam LedCount - Number of LEDs to control
+ */
+template<std::size_t LedCount>
+struct apa102_spi_frame
+{
+  // each LED needs 1 byte for bightness and 1 byte each for red, green, and
+  // blue
+  static constexpr std::size_t led_frame_byte_count = 4;
+  static constexpr std::size_t array_length = LedCount * led_frame_byte_count;
+
+  std::array<hal::byte, array_length> data;
+};
+
+struct rgb_values
+{
+  hal::byte red;
+  hal::byte green;
+  hal::byte blue;
+};
```

Comment:
I've been back and forth on this... and I think... the better way to do this is

```C++

```
```suggestion

struct apa102_pixel
{
  hal::byte brightness = 0b1111'1111;
  hal::byte red;
  hal::byte green;
  hal::byte blue;
};

static_assert(4U == sizeof(apa102_pixel), "APA102 Pixel structure must be 4 bytes in length");

/**
 * @brief Contains data to send over SPI and information about size of the data
 * to send over
 *
 * @tparam LedCount - Number of LEDs to control
 */
template<std::size_t led_count>
struct apa102_frame
{
  std::array<apa102_pixel, led_count> data;
};
```

I believe this was your original idea and I think its better than what was needed for the neopixels. The order should be consistent and in the order of the declared variables.